### ISSUE TITLE
Identifier vs ULR

### ISSUE THREAD

**Issue by rajido** [schemaorg/schemaorg#1286]
In the Life Sciences we tend to use [identifiers for our data entries](https://zenodo.org/record/18003/files/MS_2015-05-23.pdf). I imagine this might also be a common practice among many other domains.

A colleague did ask me what property to use to tag identifiers of biological datasets. I could not give a clear answer. The "[dataset](https://schema.org/Dataset)" type do not have an "identifier" property but the "URL" property could do capture this information as suggested by @proccaserra and @proccaserra in #1196 . However I do not think URL and identifier is the same thing and in some cases it might not be that easy to guess the identifier from a URL.

"4X0L" is an example of a PDB identifier and we can find several URLs (web entries) for this identifier. e.g. "https://www.ebi.ac.uk/pdbe/entry/pdb/4x0l/", "http://www.rcsb.org/pdb/explore/explore.do?structureId=4X0L", "http://identifiers.org/pdb/4x0l", ...

I have the impression at the moment each schema.org type is including an identifier property whenever it is needed. eg. "reservationId" for "reservation", "productID" for "product", ... In Bioschemas #1028 we did also have discussions about how to identify an entry. Among other things it was suggested a new "[eventId](https://docs.google.com/document/d/12O8gsOuH2qHKpis6vTHxXExraNH5XZEnoilwNgWWgaQ/edit#heading=h.2fbsxcd60sp9)" property at the level of the "event" type. The Bioschemas training materials group also suggested to create a "[pid](https://docs.google.com/document/d/1HG2fEjCoDUE4tn1XZ_ZIeWLEFXnI3YtS_FRIIFIbv-s/edit#heading=h.4p6csn8mvf4d)" property at the level of the "CreativeWork".

I think it would be worth considering to have an "identifier" property at the "thing" level instead of one specific identifier property per specific type. This is something that was suggested by @danbri #1206 and something that maybe could help to avoid some naming problems #955 .

Something else to consider is other kind of identifiers like alternative identifiers or identifiers cross references. You can find some examples in the publication referenced [above](https://zenodo.org/record/18003/files/MS_2015-05-23.pdf).

To recap, to help my colleague I was wondering if you could help me to answer the following questions ...

- Does it make sense to create a "identifier" property or should we use "URL" and "sameAs"?
- If it makes sense to use an "identifier" property should it be proposed a the level of a specific type (e.g. dataset #1083 ) or at the level of "thing"?

**Comment by jaygray0919** [comment]
We have proteins in our schema.org-enabled database. We use an Authority IRI as the `@id`. So, for the example above: "`@id`": "https://www.ebi.ac.uk/pdbe/entry/pdb/4x0l/".

Our "default" Authority in this space is EMBL. Then we carry an array of `"sameAs": [{},{},{}]` to link to the same identifier from other Authorities.

**Comment by danbri** [comment]
(ULR=URL?)

One design would be for us to add a generic textual identifier property, I'd call it something ugly like "textId" since "id" is already used in JSON-LD for @id syntax with URL values. Then we could declare all our many per-situation ID properties as subproperties of it...

**Comment by danbri** [comment]
(and then URL-valued IDs are another game entirely)

**Comment by RichardWallis** [comment]
I concur with @danbri about creating a generic identifier property that we could declare all our many per-situation ID properties as sub-properties of.

However, I still believe there is a need, in many areas, to attach more than one identifier, of differing types, to the same thing.

The introduction of a "textid"Text ranged property would not solve all these particular issues. 

I therefore suggest the proposed super-property be named something like "identifierValue" and have a range that includes Text and a new Intangible based type named "IdentifierSpecification" or "identifierDefinition" with properties: "identifierType" (Text,URL) and "identifierValue" (Text) 

Users would then have the option to either to supply a text string for an identifier or a more structured description of the type of identifier and its value.

**Comment by andrea-perego** [comment]
Another option is to define specific properties for the relevant identifier schemes. Some are already available - e.g., http://schema.org/isbn, http://schema.org/issn, http://schema.org/gtin12, etc. Similarly, others like schema:doi, schema:orcid, schema:isni, etc. can be added.

**Comment by danbri** [comment]
First draft, please take a look: http://pending.webschemas.org/identifier

We should have examples, and maybe move the explanation to supporting docs so that it can be very minimalistic since the property applies to all kinds of thing and its definition will appear all across schema.org in every page.

**Comment by danbri** [comment]
On that last point, see http://pending.webschemas.org/Thing ... all the Thing properties should have much shorter descriptions.

**Comment by RichardWallis** [comment]
Description length - tangentially related to issue (#725) .

**Comment by annaleida** [comment]
As useful as the generic identifier type is, there is (un?)fortunately several properties where identifiers would be useful, more than the one's described as sub-properties. e.g. DOI, Pubmed, ORCID, PDB, EZID etc etc, as andrea-perego pointed out. Some of them may be separated based on their regex, but for others there is a risk of confusing different identifiers with each other.

Ideally, I would like two properties for the identifier; one specifying the type of identifier, the other specifying the actual ID. This would prevent the need for a very long list of sub-properties with the need to update this. Example
IdentifierType: "ISBN"
IdentifierValue: "978-3-16-148410-0"

Another issue is that some identifiers are redirects in several layers, and here it would be useful to have robots know that the identifier stated is not the final endpoint. Ex from rajido extended:
"4X0L" is an example of a PDB identifier:
"10.2210/pdb4x0l/pdb" is the DOI for the same objects, redirecting to the PDB above.
One suggestion would be for this case to add a third property describing if the identifier is secondary or primary.

**Comment by RichardWallis** [comment]
If you checkout the proposed [identifier](http://webschemas.org/identifier) property on webschemas.org you will see that it has Text, URL and [PropertyValue](http://webschemas.org/PropertyValue) as expected types.

PropertyValue gives the option to describe the type of identifier.
eg.
```
"identifier": {
    "@type": "PropertyValue",
    "value": "0203 006 6051",
    "propertyID": "ORCID"
},
```
or 
```
"identifier": {
    "@type": "PropertyValue",
    "value": "4957203",
    "propertyID": "http://pubmed.gov/"
},
```
(not sure if that is the correct url you would use to identify a PubMed id, but it demonstrates the principle)

I didn't use [isbn](http://webschemas.org/isbn) as we already have a property specifically allocated for that.  Although being a sub-property of identifier it could also be described as above. 

As to the redirect scenario you describe, I'm not convinced how many robots would follow the indirection but you could do something like:
```
"identifier": {
    "@type": "PropertyValue",
    "description": "DOI for PDB reference",
    "propertyID": "DOI",
    "value": "10.2210/pdb4x0l/pdb",
    "value":   {
          "@type": "PropertyValue",
          "value": "4X0L",
          "propertyID": "PDB"
          }
},
```

**Comment by annaleida** [comment]
Thank you, the first two examples I think sorts that issue nicely. The last one seems a bit forced to me, but it is more because it feels a bit awkward to create 2 identifiers in the first place, but nonetheless it exists. The preferrable would of course be if it was referenced as:
"identifier":  {
          "@type": "PropertyValue",
          "value": "4X0L",
          "propertyID": "PDB"
          }

since both lead to the same resource.

**Comment by RichardWallis** [comment]
@annaleida Equally for that case you could have something like this:
```
"identifier": [
    {
       "@type": "PropertyValue",
       "propertyID": "DOI",
       "value": "10.2210/pdb4x0l/pdb"
    },
    {
          "@type": "PropertyValue",
          "value": "4X0L",
          "propertyID": "PDB"
    }
],
```

**Comment by citizencontact** [comment]
I would posit that there is a difference between a link and a citation. 
A link is a location that is the best globally unique identifier that is 
a string that can include all of the disparate parts that might be in a 
citation. I would always use a link/URL as the identifier. However, an 
electronic citation is a collection of metadata that should allow a 
person to find the resource and may or may not include a globally unique 
identifier. The best citations are a list of key pairs, many of which 
may not be in the unique identifier. For example, there may be some 
Dublin Core, FRBR and other metadata that would be helpful in a citation.

There are numerous identifier systems, like URNs, DOIs, etc. But thanks 
to the combination of a distributed domain system, the ability to find 
the resource, the option to include regex tools, the opportunity to use 
fragmentary references (for example using the id attribute with the 
#chap-subsection method).

Building the parsers for citations are difficult due to the nearly 
infinite organizations that may be publishers. Using the URL where the 
publishers of distinct repositories may expose the 
regex/listing/handle/redirect system as they choose.

In my opinion,

Daniel

**Comment by RichardWallis** [comment]
@citizencontact at this stage I don't think we are addressing the issue of detailed and accurate citation in CreativeWorks here.  

Providing a useful and consistent way of describing an identifier for any type of Thing that can be a simple string or a more complex value/identifier type pair is a significant achievable step forward, that can help the description of Products, Persons, CreativeWorks, etc.

Their introduction I don't believe would negate the future addressing the citation needs you reference.

**Comment by davux** [comment]
@rajido, I think what you want here is [JSON-LD's `@id`](https://www.w3.org/TR/json-ld/#node-identifiers). That represents a unique identifier for any object. I think it has to be a URI, but that's only a syntactic detail: if you can make up a unique identifier within its own domain field, it should be nearly trivial to word it as a URI, by prepending some identifier for the domain, pretty much how URN work.

There is not generic `id` or `identifier` in schema.org. While there could probably be, I personally think it's fair that there isn't, because a generic identifier is more a technical, "meta-property" than an actual property of whatever you're describing.
